Use livekit's Rust SDK instead of their swift SDK

maxbrunsfeld opened the pull request:
See https://github.com/livekit/rust-sdks/pull/355

Todo:

* [x] make `call` / `live_kit_client` crates use the livekit rust sdk
* [x] create a fake version of livekit rust API for integration tests
* [x] capture local audio
* [x] play remote audio
* [x] capture local video tracks
* [x] play remote video tracks
* [x] tests passing
* bugs
    * [x] deafening does not work (https://github.com/livekit/rust-sdks/issues/359)
    * [x] mute and speaking status are not replicated properly: (https://github.com/livekit/rust-sdks/issues/358)
    * [x] **linux** - crash due to symbol conflict between WebRTC's BoringSSL and libcurl's openssl (https://github.com/livekit/rust-sdks/issues/89)
    * [x] **linux** - libwebrtc-sys adds undesired dependencies on `libGL` and `libXext`
    * [x] **windows** - linker error, maybe related to the C++ stdlib (https://github.com/livekit/rust-sdks/issues/364)
        ```
        libwebrtc_sys-54978c6ad5066a35.rlib(video_frame.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in libtree_sitter_yaml-df6b0adf8f009e8f.rlib(2e40c9e35e9506f4-scanner.o)
        ```
    * [x] audio problems

Release Notes:

- Switch from Swift to Rust LiveKit SDK 🦀

maxbrunsfeld committed:
Start work on using livekit's Rust SDK instead of their swift SDK

maxbrunsfeld committed:
Initialize dispatcher for livekit

maxbrunsfeld committed:
Start implementing playing of remote audio tracks

co-authored-by: mikayla <mikayla@zed.dev>

maxbrunsfeld committed:
:art: Rename fns

maxbrunsfeld committed:
Configure audio output stream to match webrtc source stream

maxbrunsfeld committed:
Wire up screen sharing in livekit test app

maxbrunsfeld committed:
Fix some warnings

maxbrunsfeld committed:
Tweak livekit example app

maxbrunsfeld committed:
Retain screen capture sources

maxbrunsfeld committed:
Start making screen capture work

Co-authored-by: Conrad <conrad@zed.dev>

maxbrunsfeld committed:
Fix obj-c selector in MacScreenCaptureStream::drop

Co-authored-by: Mikayla <mikayla@zed.dev>

maxbrunsfeld committed:
Wire up more logic for screen sharing

Co-authored-by: Mikayla <mikayla@zed.dev>

maxbrunsfeld committed:
Use H264 codec for screen sharing

maxbrunsfeld committed:
Use RemoteVideoTrackView in SharedScreen view

maxbrunsfeld committed:
Play audio tracks that are published

maxbrunsfeld committed:
Bump livekit sha

maxbrunsfeld committed:
Fix issues with test livekit client

maxbrunsfeld committed:
Update live_kit_client crate description

maxbrunsfeld committed:
Remove todos in livekit test client

maxbrunsfeld committed:
Fix platform ScreenCaptureFrame definitions

maxbrunsfeld committed:
Conditionally compile mac-specific logic in live_kit_client

maxbrunsfeld committed:
Fix live_kit_client build errors on linux

maxbrunsfeld committed:
Merge branch 'main' into livekit-rust-sdk

maxbrunsfeld committed:
Only add the obj-c link arg on mac

maxbrunsfeld committed:
Remove spurious dependencies in non-mac builds of livekit client

maxbrunsfeld committed:
Fix some livekit integration tests

maxbrunsfeld committed:
Don't try to retrieve audio device in tests

maxbrunsfeld committed:
Remove unwraps in audio handling

maxbrunsfeld committed:
Merge branch 'main' into livekit-rust-sdk

maxbrunsfeld committed:
Fix initialization of muted state on remote participants

maxbrunsfeld committed:
Make TrackPublication::track work properly for fake livekit

maxbrunsfeld committed:
Downgrade livekit to fix audio frame queuing bug

maxbrunsfeld committed:
Add deafening to livekit test app

maxbrunsfeld committed:
Stop newly subscribed audio tracks when deafened

maxbrunsfeld committed:
Remove unwrap on test server event sending

maxbrunsfeld committed:
Make test media streams return pending

maxbrunsfeld committed:
Merge branch 'main' into livekit-rust-sdk

maxbrunsfeld committed:
Merge branch 'main' into livekit-rust-sdk

maxbrunsfeld committed:
Don't actually capture audio in tests

maxbrunsfeld committed:
Bump livekit crate for deafening fix

maxbrunsfeld committed:
Merge branch 'main' into livekit-rust-sdk

maxbrunsfeld committed:
Merge branch 'main' into livekit-rust-sdk

maxbrunsfeld committed:
Bump livekit for boringSSL symbol conflict fix

maxbrunsfeld committed:
Fix clippy warning

niklaskorz commented:
Cargo.lock contains two entries for `cpal` version `0.15.3` (upstream and the zed-fork). This causes issues at least on Nix and generally seems like something is off.

WeetHet commented:
> Cargo.lock contains two entries for `cpal` version `0.15.3` (upstream and the zed-fork). This causes issues at least on Nix and generally seems like something is off.

Yeah, it's because `rodio` also depends on `cpal`. That's easy to fix though, here's a patch

[0001-Override-cpal-to-ensure-that-the-same-version-is-use.patch](https://github.com/user-attachments/files/16644162/0001-Override-cpal-to-ensure-that-the-same-version-is-use.patch)

niklaskorz commented:
Trying to sign in with Zed using this PR branch results in a segfault:

```
Finished `dev` profile [unoptimized + debuginfo] target(s) in 1m 08s
Running `target/debug/zed`
[2024-08-20T12:05:47+02:00 ERROR fs] No such file or directory (os error 2) about ["/home/niklaskorz/.config/github-copilot"]
[2024-08-20T12:05:48+02:00 ERROR assistant] no worktrees when constructing ProjectLspAdapterDelegate
[2024-08-20T12:05:52+02:00 ERROR lsp] cannot read LSP message headers
[2024-08-20T12:05:52+02:00 ERROR lsp] Broken pipe (os error 32)
[2024-08-20T12:05:52+02:00 ERROR lsp] server shut down
Segmentation fault (core dumped)
```

WeetHet commented:
> Trying to sign in with Zed using this PR branch results in a segfault:
> 
> ```
> Finished `dev` profile [unoptimized + debuginfo] target(s) in 1m 08s
> Running `target/debug/zed`
> [2024-08-20T12:05:47+02:00 ERROR fs] No such file or directory (os error 2) about ["/home/niklaskorz/.config/github-copilot"]
> [2024-08-20T12:05:48+02:00 ERROR assistant] no worktrees when constructing ProjectLspAdapterDelegate
> [2024-08-20T12:05:52+02:00 ERROR lsp] cannot read LSP message headers
> [2024-08-20T12:05:52+02:00 ERROR lsp] Broken pipe (os error 32)
> [2024-08-20T12:05:52+02:00 ERROR lsp] server shut down
> Segmentation fault (core dumped)
> ```

Could you try running it with lldb and [sanitizers](https://doc.rust-lang.org/stable/unstable-book/compiler-flags/sanitizer.html)?

maxbrunsfeld committed:
Merge branch 'main' into livekit-rust-sdk

maxbrunsfeld committed:
Add all_load linker flag to cargo config

This ensures that objective-c categories from the webrtc-sys crate are properly included.

SomeoneToIgnore committed:
Merge branch 'main' into livekit-rust-sdk

SomeoneToIgnore committed:
Merge branch 'main' into livekit-rust-sdk

SomeoneToIgnore committed:
Disable livekit for Windows

co-authored-by: Mikayla Maki <mikayla@zed.dev>

niklaskorz commented on crates/live_kit_client/Cargo.toml:
```diff
@@ -28,28 +28,23 @@ test-support = [
-async-broadcast = "0.7"
 async-trait = { workspace = true, optional = true }
-collections = { workspace = true, optional = true }
+cpal = { git = "https://github.com/zed-industries/cpal", rev = "fd8bc2fd39f1f5fdee5a0690656caff9a26d9d50" }
```
As mentioned in https://github.com/zed-industries/zed/pull/13343#issuecomment-2294797663, the cpal dependency should be patched at the workspace level instead to avoid duplicate dependencies using the same version in the lockfile.

SomeoneToIgnore replied:
https://github.com/zed-industries/zed/pull/13343/commits/eaac4a5731be2fda7faec4a2a059150ccbc91f42

SomeoneToIgnore committed:
Merge branch 'main' into livekit-rust-sdk

SomeoneToIgnore committed:
Fix the formatting

SomeoneToIgnore committed:
Properly exclude Windows

SomeoneToIgnore committed:
Merge branch 'main' into livekit-rust-sdk

SomeoneToIgnore committed:
Use latest livekit-rust-sdk

co-authored-by: Mikayla Maki <mikayla@zed.dev>

SomeoneToIgnore committed:
Small clippy fix

mgsloan committed:
Provide HttpClient impl to our fork of LiveKit

SomeoneToIgnore committed:
Make Windows build compile

SomeoneToIgnore committed:
More fixes for the binaries to compile

SomeoneToIgnore committed:
Merge branch 'main' into livekit-rust-sdk

SomeoneToIgnore committed:
Fix other platforms

SomeoneToIgnore committed:
Test lib fixes

SomeoneToIgnore committed:
Enable the microphone in the UI

SomeoneToIgnore committed:
Move cpal override into the patch section

SomeoneToIgnore committed:
Merge branch 'main' into livekit-rust-sdk

SomeoneToIgnore committed:
Feature-gate Linux audio for now

SomeoneToIgnore committed:
Small style fix

SomeoneToIgnore committed:
Final fixes before macOS release

co-authored-by: Mikayla Maki <mikayla@zed.dev>

SomeoneToIgnore committed:
Linux: Fixed the tests, hid the screenshare

co-authored-by: Mikayla Maki <mikayla@zed.dev>

SomeoneToIgnore commented:
Small heads-up for the participants interested: we're about to merge this PR and test it internally.
This is NOT ready for Linux yet, but seems to fully replace the old macOS Swift SDK without any issues found so far.

Next rough steps look like this:
* supporting Linux audio
* fixing audio issues with wireless earphones and other dynamic frequency devices (same issue is in the old SDK, but now we can control things and fix it, finally)
* work on Linux screenshare and fixing whatever odd segfaults certain Linux distros experience

SomeoneToIgnore commented:
Not that fast, alas, we have to revert it and try harder: https://github.com/zed-industries/zed/pull/20809
